pkg/gds/secret.go

```diff
@@ -28,108 +27,180 @@ func CreateToken(length int) string {
+
+	// Build the request.
+	req := &secretmanagerpb.CreateSecretRequest{
+		Parent:   sm.parent,
+		SecretId: secretName,
+		Secret: &secretmanagerpb.Secret{
+			Replication: &secretmanagerpb.Replication{
+				Replication: &secretmanagerpb.Replication_Automatic_{
+					Automatic: &secretmanagerpb.Replication_Automatic{},
+				},
+			},
+		},
 	}
 
-	nonce, err := genNonce(nonceSize)
+	// Call the API. Note: We don't actually need the result that comes back from the API call
+	// and not accessing it directly (e.g. logging plaintext, etc) provides added security
+	_, err := sm.client.CreateSecret(sctx, req)
 	if err != nil {
-		return nil, nil, err
+		// If the API call is malformed, it will hang until the internal context times out
+		if errors.Is(err, context.DeadlineExceeded) {
+			return err
+		}
+
+		// If the secret already exists, that means the client already has a password set up
+		// This is fine because we'll just create a new version with a new password for them
+		// and CertMan will always look for the most recent secret version.
+		serr, ok := status.FromError(err)
+		if ok && serr.Code() == codes.AlreadyExists {
+			return err
+		}
+
+		// If the error is something else, something went wrong.
+		return err
 	}
+	return nil
+}
 
-	ciphertext = aesgcm.Seal(nil, nonce, []byte(plaintext), nil)
-	if len(ciphertext) == 0 {
-		return nil, nil, errors.New("could not encrypt secret with aes gcm")
+// AddSecretVersion adds a new secret version to the given secret and the
+// provided payload. Returns an error if one occurs.
+// Note: to add a secret version, the secret must first be created using CreateSecret.
+func (sm *SecretManager) AddSecretVersion(ctx context.Context, secret string, payload []byte) error {
+
+	secretPath := fmt.Sprintf("%s/secrets/%s-%s", sm.parent, sm.requestId, secret)
+	// Create an internal context, since a failed API call will result in infinite hang
+	sctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	// Build the request.
+	req := &secretmanagerpb.AddSecretVersionRequest{
+		Parent: secretPath,
+		Payload: &secretmanagerpb.SecretPayload{
+			Data: payload,
+		},
 	}
 
-	sig, err := createHMAC(key, ciphertext)
+	// Call the API. Note: We don't actually need the result that comes back from the API call
+	// and not accessing it directly (e.g. logging plaintext, etc) provides added security
+	_, err := sm.client.AddSecretVersion(sctx, req)
 	if err != nil {
-		return nil, nil, err
+		// If the API call is malformed, it will hang until the internal context times out
+		if errors.Is(err, context.DeadlineExceeded) {
+			return err
+		}
+
+		// If the secret does not exist (e.g. has been deleted or hasn't been created yet)
+		// we'll get a Not Found error
+		serr, ok := status.FromError(err)
+		if ok && serr.Code() == codes.NotFound {
+			return err
+		}
+
+		// If the error is something else, something went wrong.
+		return errors.New("unknown error: unable to create secret version")
```

Comment:
It would be preferable to pass the error up to the CertManager for logging, e.g.:

```
return fmt.Errorf("unable to create secret version: %s", err)
```